src/Agent.Worker/ICommandExtension.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker
+{
+    public interface ICommandExtension : IExtension
+    {
+        string CommandArea { get; }
+
+        void ProcessCommand(IExecutionContext context, Command command);
+    }
+
+    public class Command
+    {
+        private Dictionary<string, string> _properties;
+        private const string c_loggingEventPrefix = "##vso";
+        private static List<Tuple<String, String>> _escapeCharacters = new List<Tuple<string, string>>()
+        {
+            new Tuple<string, string>(";", "%3B"),
+            new Tuple<string, string>("\r", "%0D"),
+            new Tuple<string, string>("\n", "%0A")
+        };
+
+        public Command(string area, string eventName)
+        {
+            Area = area;
+            Event = eventName;
+        }
+
+        public string Area
+        {
+            get;
+            private set;
+        }
+
+        public string Event
+        {
+            get;
+            private set;
+        }
+
+        public Dictionary<string, string> Properties
+        {
+            get
+            {
+                if (_properties == null)
+                {
+                    _properties = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+                }
+
+                return _properties;
+            }
+        }
+
+        public string Data
+        {
+            get;
+            set;
+        }
+
+        public override String ToString()
+        {
+            StringBuilder sb = new StringBuilder();
+            sb.AppendFormat(CultureInfo.InvariantCulture, "{0}[{1}.{2}", c_loggingEventPrefix, this.Area, this.Event);
+
+            Boolean splitSpace = false;
+            foreach (var parameter in this.Properties)
+            {
+                String value = parameter.Value;
+                if (!String.IsNullOrEmpty(value))
+                {
+                    foreach (var escape in _escapeCharacters)
+                    {
+                        value = value.Replace(escape.Item1, escape.Item2);
+                    }
+
+                    if (!splitSpace)
+                    {
+                        sb.AppendFormat(CultureInfo.InvariantCulture, " ");
+                        splitSpace = true;
+                    }
+                    sb.AppendFormat(CultureInfo.InvariantCulture, "{0}={1};", parameter.Key, value);
+                }
+            }
+
+            String data = this.Data;
+            if (!String.IsNullOrEmpty(data))
+            {
+                foreach (var escape in _escapeCharacters)
+                {
+                    data = data.Replace(escape.Item1, escape.Item2);
```

Comment:
Is it just me being old fashioned or something or is the whole Item1 and Item2 thing unreadable?  Then the opposite code is replacing Item2 with Item1 ... of course! :)